packages/noports_core/lib/src/sshrvd/signature_verifying_socket_authenticator.dart

```diff
@@ -0,0 +1,118 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:at_chops/at_chops.dart';
+import 'package:at_utils/at_logger.dart';
+import 'package:socket_connector/socket_connector.dart';
+
+///
+/// Verifies signature of the data received over the socket using the same signing algorithm used to sign the data
+/// See [SigningAlgoType] to know more about supported signing algorithms
+/// See [HashingAlgoType] to know more about supported hashing algorithms
+///
+/// Expects the first message received in JSON format, with the following structure:
+/// {
+///       "signature":"<base64 encoded signature>",
+///       "hashingAlgo":"<algo>",
+///       "signingAlgo":"<algo>"
+///  }
+///
+/// also expects signature to be base64 encoded
+///
+///
+class SignatureAuthVerifier implements SocketAuthVerifier {
+  static final AtSignLogger logger = AtSignLogger('SignatureAuthVerifier');
+
+  /// Public key of the signing algorithm used to sign the data
+  String publicKey;
+
+  /// data that was signed, this is the data that should be matched once the signature is verified
+  String dataToVerify;
+
+  /// string generated by rvd which should be included in auth strings from sshnp and sshnpd
+  String rvdNonce;
+
+  /// a tag to help decipher logs
+  String tag;
+
+  SignatureAuthVerifier(
+    this.publicKey,
+    this.dataToVerify,
+    this.rvdNonce,
+    this.tag,
+  );
+
+  AtSigningResult _verifySignature(AtSigningVerificationInput input) {
+    AtChopsKeys atChopsKeys = AtChopsKeys();
+    AtChops atChops = AtChopsImpl(atChopsKeys);
+    return atChops.verify(input);
+  }
+
+  @override
+  Future<(bool, Stream<Uint8List>?)> authenticate(Socket socket) async {
+    Completer<(bool, Stream<Uint8List>?)> completer = Completer();
+    bool authenticated = false;
+    StreamController<Uint8List> sc = StreamController();
+    logger.info('SignatureAuthVerifier $tag: starting listen');
+    socket.listen((Uint8List data) {
+      if (authenticated) {
+        sc.add(data);
+      } else {
+        try {
+          final message = String.fromCharCodes(data);
```

Comment:
TODO: This code currently assumes that all of the authentication challenge response comes in a single flow. Ideally should parse the ints until we get a newline, then
- use everything up to the newline for auth
- discard the newline
- add everything after the newline to the output stream via `sc.add`